ICAFIX/hcp_fix_multi_run

```diff
@@ -103,10 +108,14 @@ opts_AddMandatory '--concat-fmri-name' 'ConcatName' 'run1_to_3' "Name (including
 
 opts_AddOptional '--motion-regression' 'doMotionRegression' 'TRUE or FALSE' "Controls whether to regress motion parameters as part of the cleaning, default FALSE" 'FALSE'
 
-opts_AddOptional '--training-file' 'TrainingData' 'somefile.RData' 'Specify a custom training file, defaults to "HCP_Style_Single_Multirun_Dedrift.RData"' ''
+opts_AddOptional '--enable-legacy-fix' 'legacyFixEnabled' 'TRUE or FALSE' "If TRUE and FSL_FIXDIR is set and contains fix, uses this fix, otherwise uses FSL_DIR/bin/fix (i.e. pyfix)" 'TRUE'
```

Comment:
Is there a particular reason we are framing this option from the context of legacy fix, rather than python fix? i.e., why not `--enable-python-fix` (or `--enable-pyfix`) instead? 
What if `TRUE` and `FSL_FIXDIR` isn't set?  Is there an informative check and error message for that situation (since "legacy" mode was requested), or could it silently resort to using `pyfix`?

Reply:
pyfix is the new default.  To use --legacy-fix you need the prior environment variable set or to use the above flag.

Reply:
That's not how the code is currently set up.  The default on `--enable-legacy-fix` (as its currently named) is `TRUE`.

Reply:
That was not what was requested either.

Reply:
It working this way was mostly my idea.  The way it works now, someone running an updated `hcp_fix_multi_run` with their existing `SetUp...` will get legacy fix, and someone running it with a fresh copy of `SetUp...` gets pyfix unless they uncomment the `FSL_FIXDIR`.  Additionally, this argument can be specified to force it to use pyfix even if `FSL_FIXDIR` is set.  This seemed like sane behavior to me, what scenario poses a problem here?

Reply:
This seems convoluted to me, and not intuitive to users.  To me `enable` implies "use"; not "permitted to use if some other variable happens to be set".  Do we have anything across the pipelines that is set in an analogous manner?

Reply:
If we didn't need to do any autodetection, that could simplify things, but it seemed like autodetection was an intended feature (hence Matt talking about `FSL_FIXDIR` being present changing the behavior).  No, I don't think we have any other autodetection, the matlab modes are probably the closest to this situation of having multiple ways to do conceptually the same thing.

Reply:
What I asked for was:
1) If FSL_FIXDIR is set as an environment variable as before, run legacy FIX without setting the argument
2) If the FSL_FIXDIR is specified with the argument, run legacy FIX (allowing the script to fully control use of legacy FIX in the future without needing the environment variable).  I didn't intend for users to both have to set an environment variable and an argument to get legacy FIX.  
3) If the environment variable is unset and this flag is not used, run pyfix (new users get pyfix by default).

Reply:
@glasserm: Can you clarify what you are thinking regarding your case (2)?  In your conceptualization, what is the argument named?
@coalsont: I think I see what you are trying to do.  I just don't think it is very intuitive.  (I certainly didn't understand why things were set up as they were prior to your explanation).

If I understand correctly, the goal is nudge people toward using `pyfix` as a new default, but make it not too difficult to continue to use matlab-based `fix`.  Given that, what if we did the following?

1. Have an argument named `--use-python-fix` that defaults to `TRUE` (or conversely, depending on how you want to frame the argument, `--use-legacy-fix` or `--use-matlab-fix` that defaults to `FALSE`).
2. The revised `SetUp` would have the line setting `FSL_FIXDIR` commented out (as is currently the case in the PR).  Given (1) and (2), someone starting completely fresh would default to using `pyfix`, which seems to be the intent (and consistent with what the current state of the code does).
3. If `FSL_FIXDIR` exists, but `--use-python-fix` = `TRUE` (the proposed default for that argument), then generate an informative error explaining the situation. This means that someone trying to run the new pipelines with an old/existing `SetUp` would get an error (which is where/how my proposal differs from what @coalsont implemented).  But I view that as acceptable and in fact appropriate here, because we WANT the user to be forced to be aware of this particular, fairly substantial change; and the fact that an implicit conflict between their previous `SetUp` and the default setting for `--use-python-fix` requires that they make a decision.
4. At that point, they can decide whether they want to use `pyfix` or matlab-based `fix`, and per our informative error message, make one simple but necessary change to proceed. e.g., Either comment out (or just remove) the setting of `FSL_FIXDIR` in their `SetUp` to switch to using `pyfix`, OR switch to using `--use-python-fix`=`FALSE` to force use of matlab-based `fix`.

Reply:
>  I didn't intend for users to both have to set an environment variable and an argument to get legacy FIX.

@glasserm That is why the default is `true`, so they don't have to set both.  My way doesn't require the user to know the path inside qunex where they chose to install fix.  As for your other two points, my way behaves as you described - users that have FSL_FIXDIR set to a legacy fix install get legacy fix, new users that copy the new `SetUp` get pyfix, neither needs to specify a new option.

Reply:
> If FSL_FIXDIR exists, but --use-python-fix = TRUE (the proposed default for that argument), then generate an informative error explaining the situation.

I don't think we want to have an error when the user tells the argument(s) that they want to run pyfix.  A main use case for the argument(s) are to allow qunex to set up legacy fix, but allow the user to choose between legacy fix or pyfix.

Reply:
As a more-explicit form that mostly follows the current logic, an option would be something like `--fix-version=<PREFER_LEGACY|PYFIX_ONLY|LEGACY_ONLY>`, with `PREFER_LEGACY` as the default so that existing setups would use legacy fix, and new setups without `FSL_FIXDIR` would use pyfix.  But, renaming the current option to something like `--detect-legacy-fix` (and rewriting the description to describe the logic, not the details) may also be an option, since I think the current logic does what is desired, and the issue is that it isn't obvious that it will fall back to pyfix.

Reply:
I think that is better, but more complicated than it needs to be, and importantly the user would still only know if `pyfix` got used despite `PREFER_LEGACY` mode by digging into the log to see what got used.

I don't understand the reluctance to generate a one-time error if trying to run `pyfix` but `FSL_FIXDIR` exists.  This is an important thing to bring to user's attention, IMO.  They then read our description of the source of the error, make a intentional decision on how they want to proceed, and then implement a one-time change to either the pyfix argument (whatever we call it), or remove the setting of `FSL_FIXDIR` from their existing `SetUp` file.  I think we are going too out of our way to hide this matlab vs. pyfix distinction.

Reply:
What the unedited (and therefore probably wrong in many respects) `SetUp...` contains is outside the scope of what a *pipeline* "prefers".  The default of the *pipeline* is to prefer to check `FSL_FIXDIR`, but if it can't find it, it will fall back to pyfix.  If both are set up (as we expect QuNex to do), then the default does in fact prefer legacy fix.  The old `SetUp...` script was unlikely to already have the correct path in it anyway.  This objection seems nonsensical to me.

Reply:
I think what QuNex does is itself somewhat "out of scope", since `HCPpipelines` itself is what we are dealing with here.  (i.e., What QuNex ends up doing to promote `pyfix` or alternatively matlab `fix` is separate from what we do there).  I don't like the notion of "falling-back" to using `pyfix` occurring in the background.  And conversely, for someone using an old `SetUp`, I can see value to making sure that they know about the new `pyfix` option, since we are presumably going to encourage people to start using that.  

I still don't understand the objection to my proposal, which I think is a clean way to handle what we are trying to accomplish, while alerting users to an important and substantial change in how FIX is implemented.  Sometimes an error is appropriate to call attention to something.  I think this is such a situation.

Reply:
QuNex isn't the only opportunity for both legacy fix and pyfix to be available (especially since pyfix is included in newer fsl), that was only an example.

The script's argument descriptions and conventions do not extend to anything outside the script, they are in terms of how that script will attempt to use its environment.

I don't know if you can convince Matt that everyone who is using QuNex (because it will have both options installed) or has legacy fix in their old `SetUp...` should all receive an error message until they set another flag (or, everyone who doesn't have it installed getting an error message instead of falling back to pyfix).  I see the argument for telling the user about this change, but I think it might be too unfriendly to break everyone's setups instead of trying to continue what they were previously doing (we make other changes to how the pipelines do processing without throwing errors in the user's face - if someone wants there to be no changes, they should stick with a single pipeline version).  If we just wanted to tell users about the existence of pyfix, an error message doesn't seem like the friendliest option (release notes are a place to list behavior changes).  We could put a warning, perhaps at the end of the log, if both fix types are present and one was selected by the script rather than the user.

In short, your suggestion seems to be "throw errors by default", which is why I object to it.  Users don't like errors, so the more errors we give them, the less users like our pipelines.  It isn't clear that this change is large enough to warrant forcing users to decide manually.

Reply:
I think how `QuNex` decides to proceed is an independent thing from what we are discussing here.  Regardless of what we do here, QuNex can be set up to either "default" to continue using matlab `fix`, or more likely, default to using `pyfix` with its internal `SetUp` appropriately constructed in such a fashion to work with `pyfix`.

Under my proposal, someone that hasn't even installed legacy fix is not going to get an error. (To do so, they would have to set `FSL_FIXDIR`, having never used legacy fix previously, which would be an odd thing to do).

My position is not that we just "throw errors by default" in all situations.  I fully agree that errors should be used sparingly. Rather, it is a judgement call on when issuing an error (i.e., halting processing) rises to the level of being appropriate.  In my opinion, *in this particular situation*, that bar is met, because I view the switch from matlab `fix` to `pyfix` as a rather **substantial internal change** -- certainly one of the largest *changes* that we've introduced in a while (as compared to just adding new pipelines and features).  

We just have different philosophical opinions on our assessment of this particular situation.  i.e., I think what we are introducing here is indeed a "large enough" charge to force users to decide manually, rather than going out of our way with something like `--fix-version=PREFER_LEGACY` (that as the default setting actually conveys a message contrary to what I presume we will promoting, both in stand-alone`HCPpipelines` installations and `QuNex`, which is that users switch to using `pyfix`).

When boiled down to that at its core, where do @glasserm and @demsarjure fall on that particular philosophical question, as applies to this particular situation?

Reply:
I struggled to follow all of that, but what I want is for people who have old FIX currently set up to keep running it until they make a change and for new users to default to pyfix.  FIX vs pyfix is not a huge change.  pyfix is allegedly a little better, but the main benefit is that it is much less painful to set up for new users.

Reply:
No it should continue to function as before, but if a user has a fresh install use pyfix instead.

Reply:
One wrinkle I see is whether "people who have old FIX currently set up" can be identified if they used QuNex to do so.  Since QuNex is responsible for the `SetUp...` stuff there, legacy fix will always be available in `FSL_FIXDIR` in all QuNex containers (unless they do something fancy).  Granted, this is just a potential consistency concern between our launchers/setup and using QuNex, so maybe it wouldn't make us reconsider the behavior we want in non-QuNex setups.

Reply:
Yes, that's why I was saying that how `QuNex` proceeds is a separate/independent thing to the discussion here.  `QuNex` won't have a concept of whether a given user "previously" used legacy fix in an earlier container, because all the `SetUp` stuff is internal to `QuNex` itself.  So, if `QuNex` sets up to use `pyfix` as its "default" (i.e., the analog to what someone will get with a fresh install in a non-QuNex setup), then a user will get `pyfix` unless they explicitly change the proposed argument AND set the location of `FSL_FIXDIR` themselves within the container.  Although, perhaps @demsarjure will add a `QuNex` specific flag/variable that when set handles both of those things (i.e., changes the argument and sets `FSL_FIXDIR`).  (Discussion of this should perhaps be moved to the `QuNex` Slack.)

Reply:
I guess another option would be that `QuNex` could internally just always set `FSL_FIXDIR` to the appropriate location within its enviroment (since it isn't something the user should ever need to have to locate themselves within the QuNex container), but then set its internal "default" for `--enable-legacy-fix` or `--fix-version` (or whatever we call it) so that `pyfix` gets used.   This would then be a case in which QuNex overrides the default for a given argument in `HCPpipelines`, BUT the functional "default" *behavior* would be the same between the two -- i.e., QuNex would default to using `pyfix`, just like a fresh install of `HCPpipelines` would end up using `pyfix`.

Reply:
How I currently have it coded on my end is that QuNex by default uses `pyfix`. If users want to use the legacy/MATLAB fix, they need to set a flag called `--hcp_legacy_fix`. Under the hood the desired behavior is achieved by adhering to the logic that is currently used in `hcp_multi_run_fix`, so I point `FSL_FIXDIR` either to the old fix (legacy behavior) or to the FSL bin dir (pyfix behavior). Do not be concerned with QuNex when it comes to how you want things implemented on your end. Like I said, if you change the logic for switching pyfix/matlab fix (e.g., through a flag or parameter), I will amend the logic on our end as well, no problems there.

I think this is the logic we want to stick with. In our tests we found out that pyfix gives better results and also does not cause as much head ache as MATLAB fix. So we want pyfix to be the default, but to allow users to use new containers on already processed data we also support the legacy option. Like already mentioned, the legacy option is currently not working due to issues with compiled MATLAB version. But once that is upgraded, things should work as planned.

Reply:
I'd be curious to see how you decided that `pyfix` gives "better results".  Was it just based on visual evaluation of the classifications?  Also, per what I wrote on Slack, it sounds like unless the QuNex folks are willing to try to compile legacy fix for the updated matlab version/MCR, then the only legacy-fix mode that will be available are matlab/octave (not compiled).